src/renderers/dom/client/ReactEventListener.js

```diff
@@ -60,29 +62,59 @@ PooledClass.addPoolingTo(
 );
 
 function handleTopLevelImpl(bookKeeping) {
-  var topLevelTarget = ReactMount.getFirstReactDOM(
-    getEventTarget(bookKeeping.nativeEvent)
-  ) || window;
-
-  // Loop through the hierarchy, in case there's any nested components.
-  // It's important that we build the array of ancestors before calling any
-  // event handlers, because event handlers can modify the DOM, leading to
-  // inconsistencies with ReactMount's node cache. See #1105.
-  var ancestor = topLevelTarget;
-  while (ancestor) {
-    bookKeeping.ancestors.push(ancestor);
-    ancestor = findParent(ancestor);
-  }
+  var path = bookKeeping.nativeEvent.path;
+  var topLevelTargetID;
+  if (path != null) {
```

Comment:
Can we split these branches into two functions? That avoids some nesting hell and also allows us to put a name to this code which is built-in documentation!